### PR TITLE
feat: [Tutorial] Using DFtoVW and exploring VW output

### PR DESCRIPTION
**Pull request by etiennekintzler**
As discussed with @JohnLangford and @jackgerrits during the last VW Open Source meeting, it would be nice to have a tutorial on how to use `DFtoVW` module to convert a dataframe into VW input format.

The PR is a first draft of this tutorial. 

I tried to emphasize on model interpretability (study of weights, distribution of predictions) and use visualizations when appropriate. I also choose real dataset (pollution in 5 chinese cities; see https://archive.ics.uci.edu/ml/datasets/PM2.5+Data+of+Five+Chinese+Cities) so the examples don't seem too contrived.

Additional points: 
- the notebook downloads the data from the web, untar it (which requires the rar tool) and writes the extracted csv files in `python/examples/PM_DATA`. This part could be improved.
- the notebook is based on the latest master version (8.11.0). If necessary I can make it work with the previous release (8.10.2).

Feel free to tell me if you see parts that are not clear or more generally if you have any suggestions !

### CODE REVIEW

**Comment by jackgerrits** [comment]
It seems like we'll release 8.11 very soon so it's a good time to put this out then!

File: python/examples/DFtoVW_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,1521 @@
...
+    "\n",
+    "We now use `DFtoVW` class to convert this dataframe to a suitable VW input format. One nice property of the class is that it will build the appropriate VW type (numerical or categorical) based on the pandas's column type.\n",
+    "\n",
+    "There are 2 usages for the class `DFtoVW`:\n",
+    "- **Basic usage**, using the `DFtoVW.from_colnames` convenience method.\n",
+    "- **Advanced usage**, that relies on the VW input format specification (see: https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format) and is build upon classes such as `Feature`, `Namespace` or `SimpleLabel`, `MulticlassLabel` etc..\n",
+    "\n",
+    "The current section illustrates the basic usage. Section 4 will present the advanced usage.\n",
+    "\n",
+    "To use the `DFtoVW.from_colnames` method: the user provides the target name as a string (parameter `y`), the features as a list of string (`x`) and the dataframe (`df`) that has to be converted to the method. He can also specify the type of label (argument `label_type`, default being `simple_label` for the regression). This call will build a `DFtoVW` object. \n",
```
Comment (olgavrou) [inline python/examples/DFtoVW_tutorial.ipynb:488]:
```suggestion
    "To use the `DFtoVW.from_colnames` method: the user provides the target name as a string (parameter `y`), the features as a list of string (`x`) and the dataframe (`df`) that has to be converted to the method. The user can also specify the type of label (argument `label_type`, default being `simple_label` for the regression). This call will build a `DFtoVW` object. \n",
```

File: python/examples/DFtoVW_tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,1521 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For categorical features, we see that VW format is `feature_name=feature_value` whereas for numerical feature the format is `feature_name:feature_value`. \n",
+    "\n",
+    "Also note that:\n",
+    "- for categorical variable, VW add `:1` behind the scene. For instance `day=14` is equivalent to `day=14:1`\n",
```
Comment (olgavrou) [inline python/examples/DFtoVW_tutorial.ipynb:554]: nit: 
```suggestion
    "- for categorical variables, VW adds `:1` behind the scene. For instance `day=14` is equivalent to `day=14:1`\n",
```

File: python/examples/DFtoVW_tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,1521 @@
...
+    "## 2.2. Define and train model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model = vw(P=1, enable_logging=True)\n",
```
Comment (olgavrou) [inline python/examples/DFtoVW_tutorial.ipynb:595]: maybe add a comment above explaining what `P=1` does?

File: python/examples/DFtoVW_tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,1521 @@
...
...JUJQ+8CqCmx8xPi+TDsjUkWILnI4zLCu5XismW09WgkutpZpqlnZjoMiJQPmwKayFEXCJy4rvvSw4GJtORomrBrLUZYw31huJ4ptiFgXuVVovOn7j5wLWN+XXC8bx3iQcH/RcG+S0bmALhSXlaPMFK2DIlWsq45x0YtGxgjmW8s0N2gX+nLqTUM5yLhoOpoAaWGorMdryco6nk9SfCbJJDxZN0yHKVMCGwu50aQCvkR9P1+5/2P44Pnrze/n1WXD3iDFOsefPPr3kCLyXZt/j1fD72Kx7cgTyX4ZecF8EgBVP6HM3kcp+ljdPOwDUKiGpnMY0wuaT+yU/3rxu1n6PR7Lr6BIFG0XSIH/yX8T//j83wEhMFqy9RHnPKeF4b57jtW649Yo4VHVQapoWw8xkKWaJ61jkBtmG0uQ0FSeUWmoGk+ZS+ZtZD/RXFhH6gQ6VYQQcdYzSjR50XdenWIoMk0WIkkm+94tKXAusBWCPS24N8mZVR3TQcLWevYHKbhAB2xdYD8xQORoYBjmUOY5ZQrLxrJYBo4TRZcorjcOpQRSKgSS7drTRRhnCdKAlI6jkeZi5RiUCbaLNM6ztYoyTVhVFqlgkis6FxHBkWWap7Oak2lKaSI6TUl1ZJLnFGnGqmlZNy3jXLOsOga5wdNhZGDbeL72mT2uqoYffTAjkZKN75hIw1ffHXFrWvDgaouWkdYKXp9V3N0reXZa0oXA19xLWdcOGwUPrrdsW8+yavmqexNePCk5nzc8XbZcbhue38upusi0MNTWMW86Yny7xPvBdcXDec040+wN0rdE81GuWdWOZ/YLtJK853jQx+cG6VtupDfPhZ8qyEd6sWmSGxhn73Iy7fii4OtjjI+FEEfAPxRCfPydd8YY440A9RkjhPjdwO8GuHfv3s/flv48sYvK7dixY8eOHb8w7ESmHV/U/LSibwmtczj/tpvpnVflnQ8cDlM+eGfMV9wbsawCSSL4yBsLPnm54dYk5d5hyek04+mi5WLd8LEnS0KEr767x7jQBNdH08pUkUpNFVpePM04yHI8gSOdMs1TfBQMU8n5tiVERyoNdycZo9Qwa2q6AJumpUgiw1Qzr1qOBjnvPy3RyhKvNUUimeaK2mrGA8nTeUuRBgqTkCqBFIJ1Zbk1zsi0p/aOItNUreeVWcfx0BBkxAcoU4NtIxvryI3EIDkpIHpY
```
Comment (olgavrou) [inline python/examples/DFtoVW_tutorial.ipynb:1245]: nit: maybe link to simple label in wiki [here](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format#simple)?

**Review by olgavrou — APPROVED** [review]
This looks great thank you for creating the tutorial! Have you tried it out on 8.10.1 and it didn't work?

File: python/examples/DFtoVW_tutorial.ipynb (hunk 4 above)
Comment (etiennekintzler) [reply python/examples/DFtoVW_tutorial.ipynb]: You meant in the place where I introduce `SimpleLabel` ? If yes I've just added a link in my last commit 6115e12 .

Line 1015 to 1018 seems to be before that but I might have misunderstood your comment.

File: python/examples/DFtoVW_tutorial.ipynb (hunk 3 above)
Comment (etiennekintzler) [reply python/examples/DFtoVW_tutorial.ipynb:602]: Yes ! done !

File: python/examples/DFtoVW_tutorial.ipynb (hunk 2 above)
Comment (etiennekintzler) [reply python/examples/DFtoVW_tutorial.ipynb]: Thank you for carefully rereading haha :)

 fixed in commit 6115e12

File: python/examples/DFtoVW_tutorial.ipynb (hunk 1 above)
Comment (etiennekintzler) [reply python/examples/DFtoVW_tutorial.ipynb]: fixed in 6115e12

**Comment by etiennekintzler** [comment]
> This looks great thank you for creating the tutorial! Have you tried it out on 8.10.1 and it didn't work?

Thank you ! It should work on version 8.10.1 but haven't tried out. I will test it.

**Comment by etiennekintzler** [comment]
> Have you tried it out on 8.10.1 and it didn't work?

**One notebook cell, where I select the attribute `name` of `Feature`** (in section 4.3) **fails on version <= 8.10.2**. This attribute was added in PR #2909 (see line https://github.com/etiennekintzler/vowpal_wabbit/blob/2f5c3c4a335892946271846e1720bfd4c8aced2b/python/vowpalwabbit/DFtoVW.py#L355 ) of the current version (8.11.0).

Except from this cell the rest of the notebook works on both version 8.10.1 and 8.10.2.

File: python/examples/DFtoVW_tutorial.ipynb (hunk 4 above)
Comment (olgavrou) [reply python/examples/DFtoVW_tutorial.ipynb]: looks good

**Comment by olgavrou** [comment]
> > Have you tried it out on 8.10.1 and it didn't work?
> 
> **One notebook cell, where I select the attribute `name` of `Feature`** (in section 4.3) **fails on version <= 8.10.2**. This attribute was added in PR #2909 (see line https://github.com/etiennekintzler/vowpal_wabbit/blob/2f5c3c4a335892946271846e1720bfd4c8aced2b/python/vowpalwabbit/DFtoVW.py#L355 ) of the current version (8.11.0).
> 
> Except from this cell the rest of the notebook works on both version 8.10.1 and 8.10.2.

Maybe before `Importing packages/modules` a small section `requirements` stating that this was developed for VW 8.11.0 but should work with older versions if the attribute `name` of `Feature` in section `4.3` isn't used?

**Comment by etiennekintzler** [comment]
> Maybe before `Importing packages/modules` a small section `requirements` stating that this was developed for VW 8.11.0 but should work with older versions if the attribute `name` of `Feature` in section `4.3` isn't used?

I've just committed the change ! 

Another possibility was to use try-catch and to use the older way to get the feature name in case `Feature.name` can't be accessed.

**Comment by jackgerrits** [comment]
@etiennekintzler FYI I'm working on docs for Python  (https://github.com/VowpalWabbit/vowpal_wabbit/pull/3093) so that this tutorial will be available on vowpalwabbit.org automatically soon

**Comment by etiennekintzler** [comment]
@jackgerrits sounds great ! you can ask me if you need any help/precisions regarding this tutorial

**Pull request merged**